source/host_utils.c

```diff
@@ -79,24 +76,38 @@ bool aws_host_utils_is_ipv6(struct aws_byte_cursor host, bool is_uri_encoded) {
     bool is_split = aws_byte_cursor_next_split(&host, '%', &substr);
     AWS_ASSERT(is_split); /* function is guaranteed to return at least one split */
 
-    if (!is_split || substr.len == 0 || s_ends_with(substr, ':') ||
-        !aws_byte_cursor_satisfies_pred(&substr, s_is_ipv6_char)) {
+    if (!is_split || substr.len < 2 || substr.len > 39 || !aws_byte_cursor_satisfies_pred(&substr, s_is_ipv6_char)) {
+        return false;
+    }
+
+    if ((substr.ptr[0] == ':' && substr.ptr[1] != ':') ||                           /* no single colon at start */
+        (substr.ptr[substr.len - 1] == ':' && substr.ptr[substr.len - 2] != ':')) { /* no single colon at end */
         return false;
     }
 
-    uint8_t group_count = 0;
+    uint8_t group_count = 1; /* string itself is the first group and then every new : we encounter is new group */
+    uint8_t digit_count = 0;
     bool has_double_colon = false;
-    struct aws_byte_cursor group = {0};
-    while (aws_byte_cursor_next_split(&substr, ':', &group)) {
-        ++group_count;
 
-        if (group_count > 8 ||                                         /* too many groups */
-            group.len > 4 ||                                           /* too many chars in group */
-            (has_double_colon && group.len == 0 && group_count > 2)) { /* only one double colon allowed */
-            return false;
+    for (size_t i = 0; i < substr.len; ++i) {
+        if (substr.ptr[i] == ':') {
+            ++group_count;
```

Comment:
Not 100% sure, but I think this will fail the legal strings "::2:3:4:5:6:7:8" and "1:2:3:4:5:6:7::" because `group_count` will be 9 at the end of this for-loop